### PR TITLE
Tweak E0277 `&`-removal suggestions

### LINKED ISSUE #64068: Diagnostics for an immutable reference to an iterator
This problem comes up in generic contexts, but the simplest code to reproduce is:

```rust
let iter = &[1,2,3].iter();
for i in iter {
}
```

Results in:

> error[E0277]: `&std::slice::Iter<'_, {integer}>` is not an iterator
> ```text
>  | for i in iter {
>  |          ^^^^ `&std::slice::Iter<'_, {integer}>` is not an iterator
>  |
> ```
>  = help: the trait `std::iter::Iterator` is not implemented for `&std::slice::Iter<'_, {integer}>`
> = note: required by `std::iter::IntoIterator::into_iter`

At first glance it looks wrong, because it says that `Iter` is not an iterator. It is not obvious that *immutability* of the shared reference is what breaks the code.

Interestingly, an only a slightly different code produces a better suggestion:

```rust
let iter = [1,2,3].iter();
for i in &iter {
}
```

> help: consider removing 1 leading `&`-references

So possibly the condition for the help text is too narrow, and could be extended to work in general case. It could also suggest using `&mut` as another alternative.

(#37914 has a similar problem, but a different root cause)

### LINKED ISSUE #84837: Confusing/incorrect "&mut &mut [T] is not an iterator"
Given the following code: 

```rust
let mut v: Vec<u8> = Vec::new();
for _ in &mut &mut v {}
```
```rust
let mut v: &mut [u8] = &mut [];
for _ in &mut v {}
```

The current output is:

```
error[E0277]: `Vec<u8>` is not an iterator
 --> src/main.rs:4:14
  |
4 |     for _ in &mut &mut v {}
  |              ^^^^^^^^^^^ `Vec<u8>` is not an iterator
  |
  = help: the trait `Iterator` is not implemented for `Vec<u8>`
  = note: required because of the requirements on the impl of `Iterator` for `&mut Vec<u8>`
  = note: 1 redundant requirements hidden
  = note: required because of the requirements on the impl of `Iterator` for `&mut &mut Vec<u8>`
  = note: required because of the requirements on the impl of `IntoIterator` for `&mut &mut Vec<u8>`
  = note: required by `into_iter`
```
```
error[E0277]: `[u8]` is not an iterator
 --> src/main.rs:3:14
  |
3 |     for _ in &mut v {}
  |              ^^^^^^ `[u8]` is not an iterator
  |
  = help: the trait `Iterator` is not implemented for `[u8]`
  = note: required because of the requirements on the impl of `Iterator` for `&mut [u8]`
  = note: 1 redundant requirements hidden
  = note: required because of the requirements on the impl of `Iterator` for `&mut &mut [u8]`
  = note: required because of the requirements on the impl of `IntoIterator` for `&mut &mut [u8]`
  = note: required by `into_iter`
  ```

Ideally it would say that `&mut &mut T` is not an iterator and _maybe_ suggest removing `&mut`.
On the other hand, `for _ in &mut &mut (0usize..)` just derefs and works fine.

@rustbot label: A-diagnostics D-incorrect D-confusing

### PR DESCRIPTION
**Pull request by estebank**
Fix #64068, fix #84837.

### CODE REVIEW

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs [hunk 1]
```diff
@@ -1302,57 +1302,117 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
...
-        let span = obligation.cause.span;
+        let mut span = obligation.cause.span;
+        let mut trait_pred = trait_pred;
+        let mut code = obligation.cause.code();
+        while let Some((c, Some(parent_trait_pred))) = code.parent() {
+            // We want the root obligation, in order to detect properly handle
+            // `for _ in &mut &mut vec![] {}`.
+            code = c;
+            trait_pred = parent_trait_pred;
+        }
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:1313]: is there some sort of peel method that does this lol

Comment (estebank) [reply compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:1313]: We have a few (blame me for the nomenclature :P) but I think not for this ^_^'

**Pull request merged**